### ISSUE TITLE
Failed to capture the lexical closure of parameter or annotation response_format

### ISSUE THREAD

**Issue by jicastillow** [MadcowD/ell#331]
Trying the example of movie reviews:

```
import ell
from pydantic import BaseModel, Field

class MovieReview(BaseModel):
    title: str = Field(description="The title of the movie")
    rating: int = Field(description="The rating of the movie out of 10")
    summary: str = Field(description="A brief summary of the movie")

@ell.complex(model="gpt-4o", response_format=MovieReview)
def generate_movie_review(movie: str):
    """You are a movie review generator. Given the name of a movie, you need to return a structured review."""
    return f"Generate a review for the movie {movie}"

review_message = generate_movie_review("The Matrix")
review = review_message.parsed
print(f"Movie: {review.title}, Rating: {review.rating}/10")
print(f"Summary: {review.summary}")
```

I get the following error:

---------------------------------------------------------------------------
```
OSError                                   Traceback (most recent call last)
File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\ell\util\closure.py:189, in _process_signature_dependency(val, dependencies, already_closed, recursion_stack, uses, name)
    [188](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:188)         else:
--> [189](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:189)             dep, _, _uses = lexical_closure(val_class, already_closed=already_closed, recursion_stack=recursion_stack.copy())
    [191](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:191) if dep: dependencies.append(dep)

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\ell\util\closure.py:86, in lexical_closure(func, already_closed, initial_call, recursion_stack, forced_dependencies)
     [84](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:84)     func = func.__ell_func__
---> [86](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:86) source = getsource(func, lstrip=True)
     [87](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:87) already_closed.add(hash(func))

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\dill\source.py:374, in getsource(object, alias, lstrip, enclosing, force, builtin)
    [373](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:373) try: # fails for builtins, and other assorted object types
--> [374](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:374)     lines, lnum = getsourcelines(object, enclosing=enclosing)
    [375](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:375) except (TypeError, IOError): # failed to get source, resort to import hooks

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\dill\source.py:345, in getsourcelines(object, lstrip, enclosing)
    [333](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:333) """Return a list of source lines and starting line number for an object.
    [334](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:334) Interactively-defined objects refer to lines in the interpreter's history.
    [335](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:335) 
   (...)
    [343](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:343) If lstrip=True, ensure there is no indentation in the first line of code.
    [344](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:344) If enclosing=True, then also return any enclosing code."""
--> [345](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:345) code, n = getblocks(object, lstrip=lstrip, enclosing=enclosing, locate=True)
    [346](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:346) return code[-1], n[-1]

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\dill\source.py:271, in getblocks(object, lstrip, enclosing, locate)
    [262](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:262) """Return a list of source lines and starting line number for an object.
    [263](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:263) Interactively-defined objects refer to lines in the interpreter's history.
    [264](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:264) 
   (...)
    [269](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:269) DEPRECATED: use 'getsourcelines' instead
    [270](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:270) """
--> [271](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:271) lines, lnum = findsource(object)
    [273](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:273) if ismodule(object):

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\dill\source.py:122, in findsource(object)
    [121](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:121) try:
--> [122](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:122)     file = getfile(object)
    [123](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/dill/source.py:123)     sourcefile = getsourcefile(object)

File C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0\Lib\inspect.py:904, in getfile(object)
    [903](file:///C:/Program%20Files/WindowsApps/PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0/Lib/inspect.py:903)     if object.__module__ == '__main__':
--> [904](file:///C:/Program%20Files/WindowsApps/PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0/Lib/inspect.py:904)         raise OSError('source code not available')
    [905](file:///C:/Program%20Files/WindowsApps/PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0/Lib/inspect.py:905) raise TypeError('{!r} is a built-in class'.format(object))

OSError: source code not available

During handling of the above exception, another exception occurred:

Exception                                 Traceback (most recent call last)
Cell In[13], [line 14](vscode-notebook-cell:?execution_count=13&line=14)
     [11](vscode-notebook-cell:?execution_count=13&line=11)     """You are a movie review generator. Given the name of a movie, you need to return a structured review."""
     [12](vscode-notebook-cell:?execution_count=13&line=12)     return f"Generate a review for the movie {movie}"
---> [14](vscode-notebook-cell:?execution_count=13&line=14) review_message = generate_movie_review("The Matrix")
     [15](vscode-notebook-cell:?execution_count=13&line=15) # review = review_message.parsed
     [16](vscode-notebook-cell:?execution_count=13&line=16) # print(f"Movie: {review.title}, Rating: {review.rating}/10")
     [17](vscode-notebook-cell:?execution_count=13&line=17) # print(f"Summary: {review.summary}")

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\ell\lmp\_track.py:131, in _track.<locals>.tracked_func(_get_invocation_id, *fn_args, **fn_kwargs)
    [126](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/lmp/_track.py:126) #XXX: cattrs add invocation origin here recursively on all pirmitive types within a message.
    [127](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/lmp/_track.py:127) #XXX: This will allow all objects to be traced automatically irrespective origin rather than relying on the API to do it, it will of vourse be expensive but unify track.
    [128](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/lmp/_track.py:128) #XXX: No other code will need to consider tracking after this point.
    [130](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/lmp/_track.py:130) if not hasattr(func_to_track, "__ell_hash__") and config.lazy_versioning:
--> [131](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/lmp/_track.py:131)     ell.util.closure.lexically_closured_source(func_to_track, forced_dependencies)
    [132](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/lmp/_track.py:132) _serialize_lmp(func_to_track)
    [134](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/lmp/_track.py:134) if not state_cache_key:

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\ell\util\closure.py:390, in lexically_closured_source(func, forced_dependencies)
    [388](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:388) if not callable(func):
    [389](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:389)     raise ValueError("Input must be a callable object (function, method, or class).")
--> [390](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:390) _, fnclosure, uses = lexical_closure(func, initial_call=True, recursion_stack=[], forced_dependencies=forced_dependencies)
    [391](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:391) return func.__ell_closure__, uses

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\ell\util\closure.py:93, in lexical_closure(func, already_closed, initial_call, recursion_stack, forced_dependencies)
     [90](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:90) dependencies, imports, modules = _process_dependencies(func, globals_and_frees, already_closed, recursion_stack, uses)
     [91](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:91) for k,v in forced_dependencies.items():
     [92](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:92)     # Todo: dictionary not necessary
---> [93](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:93)     _process_signature_dependency(v, dependencies, already_closed, recursion_stack, uses, k)
     [95](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:95) cur_src = _build_initial_source(imports, dependencies, source)
     [97](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:97) module_src = _process_modules(modules, cur_src, already_closed, recursion_stack, uses)

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\ell\util\closure.py:194, in _process_signature_dependency(val, dependencies, already_closed, recursion_stack, uses, name)
    [192](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:192)     if _uses: uses.update(_uses)
    [193](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:193) except Exception as e:
--> [194](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:194)     _raise_error(f"Failed to capture the lexical closure of parameter or annotation {name}", e, recursion_stack)

File c:\Users\JuanIgnaciodelCastil\Documents\Projects 2025\ell\.venv\Lib\site-packages\ell\util\closure.py:315, in _raise_error(message, exception, recursion_stack)
    [313](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:313) error_msg += f"Recursion stack: {' -> '.join(recursion_stack)}"
    [314](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:314) # print(error_msg)
--> [315](file:///C:/Users/JuanIgnaciodelCastil/Documents/Projects%202025/ell/.venv/Lib/site-packages/ell/util/closure.py:315) raise Exception(error_msg)

Exception: Failed to capture the lexical closure of parameter or annotation response_format. Error: source code not available
Recursion stack: generate_movie_review
```

----------------------------------------

Am i doing something wrong?

Regards.

**Comment by MadcowD** [comment]
Are you in a jupyter notebook or is there some other context?

**Comment by jicastillow** [comment]
Yes, I ran it on a Jupyter notebook.

**Comment by santiagomalter** [comment]
Same issue, with gpt-4o-2024-08-06

**Comment by asheinfeld** [comment]
Same here

**Comment by MadcowD** [comment]
Alberto are you in Jupyter?

**Comment by lanlanlanlanlanlan365** [comment]
same here.  why and how?

**Comment by chrisgoddard** [comment]
Same issue here - does Ell not support Jupyter notebooks right now?

**Comment by chrisgoddard** [comment]
If line 86 in `ell/util/closure.py` is changed to `source = getsource(func, lstrip=True, force=True)` that seems to fix the issue in jupyter

**Comment by MadcowD** [comment]
Wooo!!!!

**Comment by MadcowD** [comment]
This now is fixed with ell version 0.0.15
```
pip install --upgrade ell-ai[all]
```
